### PR TITLE
python39Packages.fontmake: init at 3.3.0

### PR DESCRIPTION
**Pull request by barinov274**
I want to add a new package. It's fontmake. Fontmake is a tool that can convert *.ufo fonts into *.ttf fonts.
Some packages were deprecated or broken that didn't allow to compile fontmake. So I changed them in separated commits.

### CODE REVIEW

File: pkgs/development/python-modules/fonttools/default.nix [hunk 1]
```diff
@@ -13,19 +13,21 @@
...
 
   disabled = pythonOlder "3.7";
 
   src = fetchFromGitHub {
     owner  = pname;
     repo   = pname;
     rev    = version;
-    sha256 = "1y9f071bdl66rsx42j16j5v53h85xra5qlg860rz3m054s2rmin9";
+    sha256 = "jTVGqNgro2QPD5x/RiR6hyYNXnJDqLnFH6RPNv0IjuE=";
+    leaveDotGit=true;
```
Comment (sternenseemann) [inline pkgs/development/python-modules/fonttools/default.nix:30]: What for? `leaveDotGit` has reproducibility problems, so I'd prefer to patch out any magic version number resolution instead (we have the information available anyways.

File: pkgs/development/python-modules/fs/default.nix [hunk 2]
```diff
@@ -32,7 +33,7 @@ buildPythonPackage rec {
   # strong cycle with paramaterized
   doCheck = false;
   checkInputs = [ pyftpdlib mock psutil pytestCheckHook ];
-  propagatedBuildInputs = [ six appdirs pytz ]
+  propagatedBuildInputs = [ six appdirs pytz setuptools ]
```
Comment (sternenseemann) [inline pkgs/development/python-modules/fs/default.nix:36]: That looks dubious, why?

**Comment by sternenseemann** [comment]
@ofborg build noto-fonts-emoji twitter-color-emoji

File: pkgs/development/python-modules/openstep-plist/default.nix [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+  pname = "openstep_plist";
+  version = "0.3.0";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "KO4sGKjod5BwUFQ1MU2S1dG0DyiJ06mdroMbRDsugBk=";
+    extension = "zip";
+  };
+
+  nativeBuildInputs = [ setuptools_scm git cython];
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/openstep-plist/default.nix:20]:
```suggestion
  nativeBuildInputs = [ setuptools-scm cython];
```
Whatever git is doing, we want to patch that. I couldn't find a usage in code on the quick.

File: pkgs/development/python-modules/openstep-plist/default.nix [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+, buildPythonPackage
+, fetchPypi
+, setuptools_scm
+, git
+, pytestCheckHook
+, cython
+}:
+
+buildPythonPackage rec {
+  pname = "openstep_plist";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/openstep-plist/default.nix:11]:
```suggestion
  pname = "openstep-plist";
```

File: pkgs/top-level/python-packages.nix [hunk 5]
```diff
@@ -3107,6 +3107,12 @@ in {
 
   fonttools = callPackage ../development/python-modules/fonttools { };
 
+  fontmake = callPackage ../development/python-modules/fontmake { };
+
+  openstep-plist = callPackage ../development/python-modules/openstep-plist { };
+
+  glyphsLib = callPackage ../development/python-modules/glyphsLib { };
```
Comment (SuperSandro2000) [inline pkgs/top-level/python-packages.nix:3114]:
```suggestion
  glyphslib = callPackage ../development/python-modules/glyphslib { };
```
or 
```suggestion
  glyphs-lib = callPackage ../development/python-modules/glyphs-lib { };
```

File: pkgs/development/python-modules/openstep-plist/default.nix [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    inherit pname version;
+    sha256 = "KO4sGKjod5BwUFQ1MU2S1dG0DyiJ06mdroMbRDsugBk=";
+    extension = "zip";
+  };
+
+  nativeBuildInputs = [ setuptools_scm git cython];
+  checkInputs = [ pytestCheckHook ];
+
+  meta = {
+    description = "A parser for the 'old style' OpenStep property list format (also known as ASCII plist), written in Cython. ";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/openstep-plist/default.nix:24]:
```suggestion
    description = "Parser for the 'old style' OpenStep property list format also known as ASCII plist";
```

File: pkgs/development/python-modules/glyphsLib/default.nix [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+  propagatedBuildInputs = [
+    fonttools
+    ufoLib2
+    openstep-plist
+    unicodedata2
+  ];
+
+  meta = {
+    description = "This Python 3.7+ library provides a bridge from Glyphs source files (.glyphs) to UFOs and Designspace files via defcon and designspaceLib.";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/glyphsLib/default.nix:35]:
```suggestion
    description = "Vridge from Glyphs source files (.glyphs) to UFOs and Designspace files via defcon and designspaceLib";
```

File: pkgs/development/python-modules/glyphsLib/default.nix [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+  format = "pyproject";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "PT66n1WEO5FNcwov8GaXT1YNrAi22X4HN7iVNkuehKI=";
+  };
+
+  checkInputs = [ pytest ];
+
+  nativeBuildInputs = [ setuptools_scm ];
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/glyphsLib/default.nix:25]:
```suggestion
  nativeBuildInputs = [ setuptools-scm ];
```

File: pkgs/development/python-modules/glyphsLib/default.nix [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+  format = "pyproject";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "PT66n1WEO5FNcwov8GaXT1YNrAi22X4HN7iVNkuehKI=";
+  };
+
+  checkInputs = [ pytest ];
+
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/glyphsLib/default.nix:24]:
```suggestion
```

File: pkgs/development/python-modules/glyphsLib/default.nix [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+  nativeBuildInputs = [ setuptools_scm ];
+
+  propagatedBuildInputs = [
+    fonttools
+    ufoLib2
+    openstep-plist
+    unicodedata2
+  ];
+
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/glyphsLib/default.nix:33]:
```suggestion

  checkInputs = [ pytestCheckHook ];

```

File: pkgs/development/python-modules/fs/default.nix (hunk 2 above)
Comment (SuperSandro2000) [reply pkgs/development/python-modules/fs/default.nix:36]: But it is correct https://github.com/PyFilesystem/pyfilesystem2/blob/master/fs/__init__.py#L4

File: pkgs/development/python-modules/fonttools/default.nix [hunk 11]
```diff
@@ -72,6 +74,7 @@ buildPythonPackage rec {
     "Tests/ufoLib"
   ];
 
+  nativeBuildInputs = [ setuptools_scm ];
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/fonttools/default.nix:77]:
```suggestion
  nativeBuildInputs = [ setuptools-scm ];
```

File: pkgs/development/python-modules/fontmake/default.nix [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+  };
+
+  propagatedBuildInputs = [
+    glyphsLib
+    ufo2ft
+    fonttools
+    fontmath
+    lxml
+  ];
+  nativeBuildInputs = [ setuptools_scm ];
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/fontmake/default.nix:29]:
```suggestion
```

File: pkgs/development/python-modules/fontmake/default.nix [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+buildPythonPackage rec {
+  pname = "fontmake";
+  version = "3.3.0";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "lD7MvZdr9CeWdoZtD3+8stVJTeQN5/AQ4miA/I2TFoE=";
+    extension = "zip";
+  };
+
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/fontmake/default.nix:21]:
```suggestion

  nativeBuildInputs = [ setuptools-scm ];

```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]
Please add `pythonImportsCheck`s to all packages and explicitly run tests.

File: pkgs/development/python-modules/fs/default.nix (hunk 2 above)
Comment (barinov274) [reply pkgs/development/python-modules/fs/default.nix:36]: when I remove "setuptools" then fontmake writes 
```
[max@nixos:~/Документы/nixpkgs]$ fontmake
Traceback (most recent call last):
  File "/nix/store/lfw6mp25fi6yk1h473z8kj9cn42wjmfg-python3.9-fontmake-3.3.0/bin/.fontmake-wrapped", line 6, in <module>
    from fontmake.__main__ import main
  File "/nix/store/lfw6mp25fi6yk1h473z8kj9cn42wjmfg-python3.9-fontmake-3.3.0/lib/python3.9/site-packages/fontmake/__main__.py", line 21, in <module>
    from ufo2ft import CFFOptimization
  File "/nix/store/m80salh8x6jrrjxl59i5m41c8mjl8734-python3.9-ufo2ft-2.27.0/lib/python3.9/site-packages/ufo2ft/__init__.py", line 14, in <module>
    from ufo2ft.outlineCompiler import OutlineOTFCompiler, OutlineTTFCompiler
  File "/nix/store/m80salh8x6jrrjxl59i5m41c8mjl8734-python3.9-ufo2ft-2.27.0/lib/python3.9/site-packages/ufo2ft/outlineCompiler.py", line 36, in <module>
    from ufo2ft.fontInfoData import (
  File "/nix/store/m80salh8x6jrrjxl59i5m41c8mjl8734-python3.9-ufo2ft-2.27.0/lib/python3.9/site-packages/ufo2ft/fontInfoData.py", line 23, in <module>
    from fontTools import ufoLib
  File "/nix/store/fmdwl461hfhj416n905qml8zhjnk3pfi-python3.9-fonttools-4.33.3/lib/python3.9/site-packages/fontTools/ufoLib/__init__.py", line 8, in <module>
    import fs
  File "/nix/store/1q6as9qzraaazq06f8x47igmsyn9dmpl-python3.9-fs-2.4.15/lib/python3.9/site-packages/fs/__init__.py", line 4, in <module>
    __import__("pkg_resources").declare_namespace(__name__)  # type: ignore
ModuleNotFoundError: No module named 'pkg_resources'
```

File: pkgs/development/python-modules/openstep-plist/default.nix [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+  version = "0.3.0";
+
+  src = fetchPypi {
+    pname = "openstep_plist";
+    inherit version;
+    sha256 = "KO4sGKjod5BwUFQ1MU2S1dG0DyiJ06mdroMbRDsugBk=";
+    extension = "zip";
+  };
+
+  nativeBuildInputs = [ setuptools-scm cython];
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/openstep-plist/default.nix:21]:
```suggestion
  nativeBuildInputs = [ setuptools-scm cython ];
```

File: pkgs/development/python-modules/glyphslib/default.nix [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+  propagatedBuildInputs = [
+    fonttools
+    openstep-plist
+    ufoLib2
+    unicodedata2
+  ];
+
+  doCheck = false;
+
+  pythonImportsCheck = [ "fontTools" "ufoLib2" "openstep_plist" "unicodedata" ];
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/glyphslib/default.nix:37]:
```suggestion
  pythonImportsCheck = [ "glyphsLib" ];
```

File: pkgs/development/python-modules/openstep-plist/default.nix [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+  checkInputs = [ pytestCheckHook ];
+
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/openstep-plist/default.nix:23]: pythonImportsCheck missing

File: pkgs/development/python-modules/fonttools/default.nix [hunk 17]
```diff
@@ -72,6 +73,7 @@ buildPythonPackage rec {
     "Tests/ufoLib"
   ];
 
+  nativeBuildInputs = [ setuptools-scm ];
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/fonttools/default.nix:77]: pythonImportsCheck missing and nativeBuildInputs goes after src and before checkInputs.

File: pkgs/development/python-modules/fontmake/default.nix [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+  propagatedBuildInputs = [
+    glyphslib
+    ufo2ft
+    fonttools
+    fontmath
+    lxml
+    setuptools
+  ];
+  nativeBuildInputs = [ setuptools-scm ];
+
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/fontmake/default.nix:32]: pythonImportsCheck missing

File: pkgs/development/python-modules/fontmake/default.nix [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+  meta = {
+    description = "fontmake compiles fonts from various sources (.glyphs, .ufo, designspace) into binaries (.otf, .ttf).";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/fontmake/default.nix:34]:
```suggestion
    description = "Compiles fonts from various sources (.glyphs, .ufo, designspace) into binaries formats (.otf, .ttf)";
```

File: pkgs/development/python-modules/glyphslib/default.nix [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+  propagatedBuildInputs = [
+    fonttools
+    openstep-plist
+    ufoLib2
+    unicodedata2
+  ];
+
+  doCheck = false;
+
+  pythonImportsCheck = [ "glyphsLib" ];
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/glyphslib/default.nix:37]:
```suggestion
  nativeBuildInputs = [ setuptools-scm ];

  propagatedBuildInputs = [
    fonttools
    openstep-plist
    ufoLib2
    unicodedata2
  ];

  checkInputs = [ pytestCheckHook ];

  pythonImportsCheck = [ "glyphsLib" ];
```

If we stick to the standard ordering it makes everything easier and you need to think and search less when debugging.

Also do we really want to disable tests here? If so why? We always want to add a comment.

File: pkgs/development/python-modules/fonttools/default.nix [hunk 21]
```diff
@@ -13,21 +13,25 @@
...
   src = fetchFromGitHub {
     owner  = pname;
     repo   = pname;
     rev    = version;
-    sha256 = "1y9f071bdl66rsx42j16j5v53h85xra5qlg860rz3m054s2rmin9";
+    sha256 = "MUIZGnYwlfTat9655AOYgK5r6AvHj/xXghUvOZR8HIM=";
   };
 
+  nativeBuildInputs = [ setuptools-scm ];
+  pythonImportsCheck = [ "fontTools" ];
```
[the remark this answers is not in the document]
Comment (lovek323) [reply pkgs/development/python-modules/fonttools/default.nix:33]: Out of interest, are these conventions documented anywhere?

File: pkgs/development/python-modules/glyphslib/default.nix [hunk 22]
```diff
@@ -0,0 +1,47 @@
...
+
+  format = "wheel";
+
+  src = fetchPypi{
+    pname = "glyphsLib";
+    inherit version format;
+    dist = "py3";
+    python = "py3";
+    sha256 = "zoiNIISxdJIjQQncRse1tJ3Gg+9Tgku8K+GUzQ8w8iQ=";
+  };
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/glyphslib/default.nix:25]: We could have merged this if you didn't change this to wheel. We where building from source before which is always preferable. Why do we want to patch the wheel if we can build it?

File: pkgs/development/python-modules/ufoLib2/default.nix [hunk 23]
```diff
@@ -25,6 +26,8 @@ buildPythonPackage rec {
     fs
   ];
 
+  nativeBuildInputs = [ setuptools-scm ];
```
Comment (jtojnar) [inline pkgs/development/python-modules/ufoLib2/default.nix:29]: Would be nice to remain consistent regarding the formatting inside each file (one dependency per line for easier `git blame`).

File: pkgs/development/python-modules/skia-pathops/default.nix [hunk 24]
```diff
@@ -0,0 +1,42 @@
...
+, pytestCheckHook
+}:
+
+buildPythonPackage rec {
+  pname = "skia-pathops";
+  version = "0.7.2";
+
+  #skia-pathops building from source tries to build "skia" for first and it causes an error "urllib.error.URLError: <urlopen error [Errno -3] Temporary failure in name resolution>"
+  #This is why we use wheel.
+  format = "wheel";
```
[the remark this answers is not in the document]
Comment (SuperSandro2000) [reply pkgs/development/python-modules/skia-pathops/default.nix:16]: Take this

```diff
diff --git a/pkgs/development/python-modules/skia-pathops/default.nix b/pkgs/development/python-modules/skia-pathops/default.nix
index e0a8f21a15a..c0ce6853e33 100644
--- a/pkgs/development/python-modules/skia-pathops/default.nix
+++ b/pkgs/development/python-modules/skia-pathops/default.nix
@@ -1,9 +1,12 @@
 { lib
+, stdenv
 , buildPythonPackage
 , fetchPypi
-, autoPatchelfHook
+, cython
+, gn
+, ninja
 , setuptools
-, stdenv
+, setuptools-scm
 , pytestCheckHook
 }:
 
@@ -11,25 +14,24 @@ buildPythonPackage rec {
   pname = "skia-pathops";
   version = "0.7.2";
 
-  #skia-pathops building from source tries to build "skia" for first and it causes an error "urllib.error.URLError: <urlopen error [Errno -3] Temporary failure in name resolution>"
-  #This is why we use wheel.
-  format = "wheel";
-
-  src = fetchPypi{
-    pname = "skia_pathops";
-    inherit version format;
-    dist = "cp39";
-    python = "cp39";
-    abi = "cp39";
-    platform = "manylinux_2_17_x86_64.manylinux2014_x86_64";
-    sha256 = "GTKjh7FMlmH0pXjP10pxsBQ2rcxO89D1JrhbfT8o8qM=";
+  src = fetchPypi {
+    pname = "skia-pathops";
+    inherit version;
+    extension = "zip";
+    sha256 = "sha256-Gdhcmv77oVr5KxPIiJlk935jgvWPQsYEC0AZ6yjLppA=";
   };
 
-  buildInputs = [ stdenv.cc.cc.lib ];
+  postPatch = ''
+    substituteInPlace setup.py \
+      --replace "build_cmd = [sys.executable, build_skia_py, build_dir]" \
+        'build_cmd = [sys.executable, build_skia_py, "--no-fetch-gn", "--no-virtualenv", "--gn-path", "${gn}/bin/gn", build_dir]'
+  '';
+
+  nativeBuildInputs = [ cython ninja setuptools-scm ];
 
   propagatedBuildInputs = [ setuptools ];
 
-  nativeBuildInputs = [ autoPatchelfHook ];
+  checkInputs = [ pytestCheckHook ];
 
   pythonImportsCheck = [ "pathops" ];
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/development/python-modules/skia-pathops/default.nix [hunk 25]
```diff
@@ -0,0 +1,42 @@
...
+
+  src = fetchPypi{
+    pname = "skia_pathops";
+    inherit version format;
+    dist = "cp39";
+    python = "cp39";
+    abi = "cp39";
+    platform = "manylinux_2_17_x86_64.manylinux2014_x86_64";
+    sha256 = "GTKjh7FMlmH0pXjP10pxsBQ2rcxO89D1JrhbfT8o8qM=";
```
[the remark this answers is not in the document]
Comment (jtojnar) [reply pkgs/development/python-modules/skia-pathops/default.nix:25]: Why would you want to include the hash type twice?

[the remark this answers is not in the document]
Comment (lovek323) [reply pkgs/development/python-modules/skia-pathops/default.nix:25]: The `sha256` property has been dropped and replaced with the `hash` property – I assume this is a new way of handling multiple kinds of hashes?

[the remark this answers is not in the document]
Comment (SuperSandro2000) [reply pkgs/development/python-modules/skia-pathops/default.nix]: Yes and not really at this point.

**Pull request merged**